Store element offsets, for Node::child_at_offset

CAD97 opened the pull request:
Sorbus needs some tests to be run under miri before I'm completely confident in this implementation, but I'm fairly certain it's correct. We don't explicitly take advantage of the fact that the order always flip-flops, and rely on the compiler to notice such to eliminate alignment checks.

_Once we have tests/benchmarks_, we can add parity tracking to `green::node::Children` such that we can unroll internal iteration to size-2 strides that do not have to check parity. `Drop` already does this, it's just more work for `Children` which also does external iteration.

Additionally, since `Element` is now an alignment-sensitive type, more of the construction code is dealing in `NodeOrToken<Arc<Node>, Arc<Token>>` than previously. We can probably move the compression/erasure to `Union2<Arc<Node>, Arc<Token>>` earlier to reduce the amount of data said functions have to move around.

-----

cc @matklad

CAD97 committed:
Remove impl details of Element

CAD97 committed:
Implement odd stride Element

CAD97 committed:
Node children iterator

CAD97 committed:
Node construction

CAD97 committed:
Remove allow(unused)

CAD97 committed:
Actually drop nodes' children elements

CAD97 committed:
Add missing TextLen for &Token

CAD97 committed:
Informally mark public dependencies

CAD97 committed:
Note on hashbrown dependency

CAD97 committed:
Add simple smoke test

CAD97 committed:
Improve debug format

And use that to actually test in the test

CAD97 committed:
Add missing green::Builder::new

CAD97 committed:
Add a second test

CAD97 committed:
Improve UB-defense debug_assert! messages

CAD97 committed:
Add simple benchmark of children iteration/visiting

CAD97 commented:
I implemented alignment tracking in `Children`, and I got an interesting result.

<details><summary>Diff</summary>

```diff
diff --git a/src/green/node.rs b/src/green/node.rs
index c1e6338..6bcf8f9 100644
--- a/src/green/node.rs
+++ b/src/green/node.rs
@@ -71,7 +71,7 @@ impl Node {
 
     /// Child elements of this node.
     pub fn children(&self) -> Children<'_> {
-        Children { inner: self.children.iter() }
+        Children { inner: self.children.iter(), full_align: true }
     }
 
     /// Child element containing the given offset from the start of this node.
@@ -214,19 +214,18 @@ unsafe impl SliceDst for Node {
 #[derive(Debug, Clone)]
 pub struct Children<'a> {
     inner: slice::Iter<'a, Element>,
-    // NB: Children can (and probably should) keep track of the alignment of
-    // the inner slice iterator. That way, the compiler should be able to pick
-    // up on the exclusively flip-flop pattern and optimize out checking for
-    // alignment. This is most important for internal iteration (fold), where
-    // we can guarantee this by unrolling to a two-stride manually. I've just
-    // not done this yet to avoid the extra safety-critical code initially.
+    full_align: bool,
 }
 
 impl<'a> Children<'a> {
     /// Get the next item in the iterator without advancing it.
     pub fn peek(&self) -> Option<NodeOrToken<ArcBorrow<'a, Node>, ArcBorrow<'a, Token>>> {
         let element = self.inner.as_slice().first()?;
-        Some(element.into())
+        if self.full_align {
+            unsafe { Some(element.full_aligned().into()) }
+        } else {
+            unsafe { Some(element.half_aligned().into()) }
+        }
     }
 
     /// Get the nth item in the iterator without advancing it.
@@ -235,7 +234,12 @@ impl<'a> Children<'a> {
         n: usize,
     ) -> Option<NodeOrToken<ArcBorrow<'a, Node>, ArcBorrow<'a, Token>>> {
         let element = self.inner.as_slice().get(n)?;
-        Some(element.into())
+        let full_align = self.full_align ^ (n % 2 == 1);
+        if full_align {
+            unsafe { Some(element.full_aligned().into()) }
+        } else {
+            unsafe { Some(element.half_aligned().into()) }
+        }
     }
 
     /// Divide this iterator into two at an index.
@@ -248,7 +252,12 @@ impl<'a> Children<'a> {
     /// Panics if `mid > len`.
     pub fn split_at(&self, mid: usize) -> (Self, Self) {
         let (left, right) = self.inner.as_slice().split_at(mid);
-        (Children { inner: left.iter() }, Children { inner: right.iter() })
+        let left_full_align = self.full_align;
+        let right_full_align = self.full_align ^ (mid % 2 == 1);
+        (
+            Children { inner: left.iter(), full_align: left_full_align },
+            Children { inner: right.iter(), full_align: right_full_align },
+        )
     }
 }
 
@@ -257,7 +266,13 @@ impl<'a> Iterator for Children<'a> {
 
     fn next(&mut self) -> Option<Self::Item> {
         let element = self.inner.next()?;
-        Some(element.into())
+        let full_align = self.full_align;
+        self.full_align = !full_align;
+        if full_align {
+            unsafe { Some(element.full_aligned().into()) }
+        } else {
+            unsafe { Some(element.half_aligned().into()) }
+        }
     }
 
     fn size_hint(&self) -> (usize, Option<usize>) {
@@ -274,7 +289,46 @@ impl<'a> Iterator for Children<'a> {
 
     fn nth(&mut self, n: usize) -> Option<Self::Item> {
         let element = self.inner.nth(n)?;
-        Some(element.into())
+        let full_align = self.full_align ^ (n % 2 == 1);
+        self.full_align = !full_align;
+        if full_align {
+            unsafe { Some(element.full_aligned().into()) }
+        } else {
+            unsafe { Some(element.half_aligned().into()) }
+        }
+    }
+
+    fn fold<B, F>(mut self, init: B, mut f: F) -> B
+    where
+        F: FnMut(B, Self::Item) -> B,
+    {
+        let mut accum = init;
+
+        macro_rules! next {
+            ($aligned:ident) => {
+                if let Some(element) = self.inner.next() {
+                    unsafe { element.$aligned().into() }
+                } else {
+                    break accum;
+                }
+            };
+        }
+
+        if self.full_align {
+            loop {
+                let el = next!(full_aligned);
+                accum = f(accum, el);
+                let el = next!(half_aligned);
+                accum = f(accum, el);
+            }
+        } else {
+            loop {
+                let el = next!(half_aligned);
+                accum = f(accum, el);
+                let el = next!(full_aligned);
+                accum = f(accum, el);
+            }
+        }
     }
 }
 
@@ -288,12 +342,26 @@ impl ExactSizeIterator for Children<'_> {
 impl DoubleEndedIterator for Children<'_> {
     fn next_back(&mut self) -> Option<Self::Item> {
         let element = self.inner.next_back()?;
-        Some(element.into())
+        // self.len() is now the index of the element popped from the back
+        let full_align = self.full_align ^ (self.len() % 2 == 1);
+        // don't change self.full_align, the alignment of the head
+        if full_align {
+            unsafe { Some(element.full_aligned().into()) }
+        } else {
+            unsafe { Some(element.half_aligned().into()) }
+        }
     }
 
     fn nth_back(&mut self, n: usize) -> Option<Self::Item> {
         let element = self.inner.nth_back(n)?;
-        Some(element.into())
+        // self.len() is now the index of the element popped from the back
+        let full_align = self.full_align ^ (self.len() % 2 == 1);
+        // don't change self.full_align, the alignment of the head
+        if full_align {
+            unsafe { Some(element.full_aligned().into()) }
+        } else {
+            unsafe { Some(element.half_aligned().into()) }
+        }
     }
 }
```
</details>

Specifically, performance on the [flat children iteration](https://github.com/CAD97/sorbus/blob/15894fc1ae4f2704d16b2617798332fc572005ed/benches/node_children.rs#L49-L60) regressed by about 9% in time (8% in throughput), but the [tree visitor benchmark](https://github.com/CAD97/sorbus/blob/15894fc1ae4f2704d16b2617798332fc572005ed/benches/node_children.rs#L69-L80) saw an improvement of about 33% in time (50%(!) in throughput).

I don't know how to interpret these results, so I'm just going to put the diff in this comment for now. @matklad, if you've got a bit of time to spare (the benchmark suite is rather quick, just a couple minutes) I'd like to see if you can reproduce these results. (Do `cargo bench` on 15894fc, and then on the same commit with the above diff applied.)

CAD97 commented:
I redid the benchmark with a larger range of input sizes, from small (2) through the big (4096) that I was doing previously. The regression percentage of the flat test grew for small numbers, but the improvement percentage of the visit test stayed consistent. I've attached the criterion report here. [criterion.zip](https://github.com/CAD97/sorbus/files/4465767/criterion.zip)

The fact that the visit performance diff is consistent makes me more prone to believe it than the flat iteration. I'll see if I can't figure out how to throw a profiler/flamegraph at it to see where work is being done in both benchmarks, but I'd also appreciate if anybody comes up with ideas on how to diagnose why these benchmarks are going in opposite directions.

CAD97 commented:
I cannot figure out profiling Rust on Windows 😞

lnicola commented:
I don't think these really help (`rayon`, really?), but I tried to run your benchmarks on my laptop:

Baseline:

```
flat_children/256       time:   [2.2180 us 2.2187 us 2.2194 us]                               
                        thrpt:  [115.35 Melem/s 115.38 Melem/s 115.42 Melem/s]
Found 13 outliers among 100 measurements (13.00%)
  6 (6.00%) high mild
  7 (7.00%) high severe
flat_children/512       time:   [4.4331 us 4.4345 us 4.4360 us]                               
                        thrpt:  [115.42 Melem/s 115.46 Melem/s 115.49 Melem/s]
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) high mild
  9 (9.00%) high severe
flat_children/1024      time:   [8.8643 us 8.8667 us 8.8694 us]                                
                        thrpt:  [115.45 Melem/s 115.49 Melem/s 115.52 Melem/s]
Found 11 outliers among 100 measurements (11.00%)
  3 (3.00%) high mild
  8 (8.00%) high severe
flat_children/2048      time:   [17.742 us 17.750 us 17.763 us]                                
                        thrpt:  [115.30 Melem/s 115.38 Melem/s 115.43 Melem/s]
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) high mild
  7 (7.00%) high severe
flat_children/4096      time:   [35.474 us 35.488 us 35.506 us]                                
                        thrpt:  [115.36 Melem/s 115.42 Melem/s 115.47 Melem/s]
Found 15 outliers among 100 measurements (15.00%)
  2 (2.00%) high mild
  13 (13.00%) high severe

visit_children/256      time:   [55.097 us 55.125 us 55.162 us]                               
                        thrpt:  [4.6408 Melem/s 4.6440 Melem/s 4.6464 Melem/s]
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) high mild
  8 (8.00%) high severe
visit_children/512      time:   [110.14 us 110.19 us 110.26 us]                               
                        thrpt:  [4.6434 Melem/s 4.6465 Melem/s 4.6487 Melem/s]
Found 14 outliers among 100 measurements (14.00%)
  2 (2.00%) high mild
  12 (12.00%) high severe
visit_children/1024     time:   [220.45 us 220.65 us 220.90 us]                                
                        thrpt:  [4.6355 Melem/s 4.6409 Melem/s 4.6450 Melem/s]
Found 17 outliers among 100 measurements (17.00%)
  8 (8.00%) high mild
  9 (9.00%) high severe
visit_children/2048     time:   [440.84 us 441.04 us 441.33 us]                                
                        thrpt:  [4.6405 Melem/s 4.6436 Melem/s 4.6457 Melem/s]
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) high mild
  10 (10.00%) high severe
visit_children/4096     time:   [881.47 us 882.01 us 882.85 us]                                
                        thrpt:  [4.6395 Melem/s 4.6440 Melem/s 4.6468 Melem/s]
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) high mild
  10 (10.00%) high severe
```

Patch:

```
flat_children/256       time:   [2.2880 us 2.2893 us 2.2912 us]                               
                        thrpt:  [111.73 Melem/s 111.82 Melem/s 111.89 Melem/s]
                 change:
                        time:   [+2.8837% +3.1852% +3.4741%] (p = 0.00 < 0.05)
                        thrpt:  [-3.3575% -3.0869% -2.8029%]
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) high mild
  8 (8.00%) high severe
flat_children/512       time:   [4.5608 us 4.5677 us 4.5773 us]                               
                        thrpt:  [111.86 Melem/s 112.09 Melem/s 112.26 Melem/s]
                 change:
                        time:   [+2.6174% +2.9550% +3.3091%] (p = 0.00 < 0.05)
                        thrpt:  [-3.2031% -2.8702% -2.5506%]
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) high mild
  8 (8.00%) high severe
flat_children/1024      time:   [9.1125 us 9.1147 us 9.1171 us]                                
                        thrpt:  [112.32 Melem/s 112.35 Melem/s 112.37 Melem/s]
                 change:
                        time:   [+2.4915% +2.8158% +3.1360%] (p = 0.00 < 0.05)
                        thrpt:  [-3.0406% -2.7387% -2.4309%]
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) high mild
  8 (8.00%) high severe
flat_children/2048      time:   [18.215 us 18.238 us 18.272 us]                                
                        thrpt:  [112.09 Melem/s 112.29 Melem/s 112.43 Melem/s]
                 change:
                        time:   [+2.5787% +2.8947% +3.2708%] (p = 0.00 < 0.05)
                        thrpt:  [-3.1672% -2.8132% -2.5138%]
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) high mild
  7 (7.00%) high severe
flat_children/4096      time:   [36.468 us 36.481 us 36.494 us]                                
                        thrpt:  [112.24 Melem/s 112.28 Melem/s 112.32 Melem/s]
                 change:
                        time:   [+2.5433% +2.7831% +3.0114%] (p = 0.00 < 0.05)
                        thrpt:  [-2.9233% -2.7077% -2.4803%]
                        Performance has regressed.
Found 15 outliers among 100 measurements (15.00%)
  7 (7.00%) high mild
  8 (8.00%) high severe

visit_children/256      time:   [31.412 us 31.459 us 31.525 us]                                
                        thrpt:  [8.1206 Melem/s 8.1377 Melem/s 8.1498 Melem/s]
                 change:
                        time:   [-43.075% -42.920% -42.762%] (p = 0.00 < 0.05)
                        thrpt:  [+74.710% +75.194% +75.668%]
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  7 (7.00%) high mild
  8 (8.00%) high severe
visit_children/512      time:   [62.771 us 62.798 us 62.829 us]                               
                        thrpt:  [8.1491 Melem/s 8.1531 Melem/s 8.1566 Melem/s]
                 change:
                        time:   [-43.161% -42.975% -42.793%] (p = 0.00 < 0.05)
                        thrpt:  [+74.803% +75.361% +75.935%]
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) high mild
  8 (8.00%) high severe
visit_children/1024     time:   [125.92 us 126.21 us 126.60 us]                                
                        thrpt:  [8.0886 Melem/s 8.1138 Melem/s 8.1323 Melem/s]
                 change:
                        time:   [-43.072% -42.864% -42.622%] (p = 0.00 < 0.05)
                        thrpt:  [+74.283% +75.021% +75.659%]
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  4 (4.00%) high mild
  10 (10.00%) high severe
visit_children/2048     time:   [251.59 us 251.76 us 251.93 us]                                
                        thrpt:  [8.1292 Melem/s 8.1349 Melem/s 8.1403 Melem/s]
                 change:
                        time:   [-43.079% -42.894% -42.666%] (p = 0.00 < 0.05)
                        thrpt:  [+74.418% +75.112% +75.684%]
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe
visit_children/4096     time:   [505.36 us 505.81 us 506.39 us]                                
                        thrpt:  [8.0887 Melem/s 8.0980 Melem/s 8.1051 Melem/s]
                 change:
                        time:   [-42.832% -42.669% -42.516%] (p = 0.00 < 0.05)
                        thrpt:  [+73.962% +74.426% +74.922%]
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) high mild
  10 (10.00%) high severe
```

`perf difff`

```
# Event 'cycles:u'
#
# Baseline  Delta Abs  Shared Object                   Symbol                                                                                                                                                                                                                                         
# ........  .........  ..............................  ...............................................................................................................................................................................................................................................
#
              +27.50%  node_children-29d7c4770dadb48b  [.] <sorbus::green::node::Children as core::iter::traits::iterator::Iterator>::fold
     0.00%    +20.67%  node_children-29d7c4770dadb48b  [.] rayon::iter::plumbing::bridge_producer_consumer::helper
               +8.05%  node_children-29d7c4770dadb48b  [.] sorbus::green::element::<impl core::convert::From<&sorbus::green::element::FullAlignedElement> for sorbus::utils::NodeOrToken<rc_borrow::ArcBorrow<sorbus::green::node::Node>,rc_borrow::ArcBorrow<sorbus::green::token::Token>>>::from
               +6.36%  node_children-29d7c4770dadb48b  [.] sorbus::green::element::<impl core::convert::From<&sorbus::green::element::HalfAlignedElement> for sorbus::utils::NodeOrToken<rc_borrow::ArcBorrow<sorbus::green::node::Node>,rc_borrow::ArcBorrow<sorbus::green::token::Token>>>::from
               +4.08%  node_children-29d7c4770dadb48b  [.] ptr_union::check_tag
    10.14%     +3.38%  libm-2.31.so                    [.] 0x000000000000f350
     5.26%     +2.82%  node_children-29d7c4770dadb48b  [.] rayon::slice::quicksort::recurse
               +1.84%  node_children-29d7c4770dadb48b  [.] oorandom::Rand64::rand_range
               +1.80%  node_children-29d7c4770dadb48b  [.] sorbus::green::element::Element::full_aligned
               +1.54%  node_children-29d7c4770dadb48b  [.] ptr_union::unset_tag
     3.00%     +0.87%  libm-2.31.so                    [.] exp
               +0.76%  node_children-29d7c4770dadb48b  [.] criterion::stats::univariate::resamples::Resamples<A>::next
               +0.32%  node_children-29d7c4770dadb48b  [.] core::ops::function::impls::<impl core::ops::function::Fn<A> for &F>::call
     0.01%     +0.29%  node_children-29d7c4770dadb48b  [.] criterion::stats::bivariate::resamples::Resamples<X,Y>::next
               +0.27%  node_children-29d7c4770dadb48b  [.] sorbus::green::node::Node::children
     0.20%     -0.19%  node_children-29d7c4770dadb48b  [.] rayon::slice::quicksort::partial_insertion_sort
     0.00%     +0.14%  node_children-29d7c4770dadb48b  [.] criterion::analysis::estimates::stats
     0.14%     -0.13%  node_children-29d7c4770dadb48b  [.] criterion::stats::univariate::sample::Sample<A>::percentiles
     0.01%     +0.12%  node_children-29d7c4770dadb48b  [.] criterion::analysis::compare::estimates::stats
     0.00%     +0.11%  node_children-29d7c4770dadb48b  [.] criterion::stats::dot
     0.09%     +0.05%  libc-2.31.so                    [.] __memmove_avx_unaligned_erms
     0.03%     -0.03%  node_children-29d7c4770dadb48b  [.] criterion::plot::plotters_backend::distributions::rel_distributions
               +0.03%  node_children-29d7c4770dadb48b  [.] crossbeam_epoch::default::pin
               +0.03%  node_children-29d7c4770dadb48b  [.] criterion::Bencher<M>::iter
     0.11%     -0.02%  node_children-29d7c4770dadb48b  [.] criterion::stats::univariate::sample::Sample<A>::median_abs_dev
     0.03%     +0.02%  node_children-29d7c4770dadb48b  [.] criterion::kde::sweep_and_estimate
               +0.02%  node_children-29d7c4770dadb48b  [.] crossbeam_deque::Stealer<T>::steal
               +0.02%  node_children-29d7c4770dadb48b  [.] <core::iter::adapters::chain::Chain<A,B> as core::iter::traits::iterator::Iterator>::try_fold
     0.02%     -0.01%  node_children-29d7c4770dadb48b  [.] rayon_core::registry::WorkerThread::wait_until_cold
     0.00%     +0.01%  node_children-29d7c4770dadb48b  [.] crossbeam_queue::seg_queue::SegQueue<T>::pop
     0.00%     +0.01%  node_children-29d7c4770dadb48b  [.] rayon::slice::quicksort::break_patterns
               +0.01%  node_children-29d7c4770dadb48b  [.] core::slice::<impl [T]>::copy_from_slice
               +0.01%  node_children-29d7c4770dadb48b  [.] crossbeam_epoch::default::HANDLE::__getit
     0.00%     +0.00%  node_children-29d7c4770dadb48b  [.] alloc::raw_vec::RawVec<T,A>::reserve
     0.01%     +0.00%  libc-2.31.so                    [.] cfree@GLIBC_2.2.5
               +0.00%  node_children-29d7c4770dadb48b  [.] <(alloc::vec::Vec<A>, alloc::vec::Vec<B>, alloc::vec::Vec<C>, alloc::vec::Vec<D>) as criterion::stats::tuple::TupledDistributionsBuilder>::push
               +0.00%  node_children-29d7c4770dadb48b  [.] alloc::vec::Vec<T>::into_boxed_slice
               +0.00%  node_children-29d7c4770dadb48b  [.] crossbeam_deque::Worker<T>::pop
               +0.00%  node_children-29d7c4770dadb48b  [.] <crossbeam_epoch::sync::list::Iter<T,C> as core::iter::traits::iterator::Iterator>::next
     0.01%     -0.00%  node_children-29d7c4770dadb48b  [.] <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
     0.02%     +0.00%  [unknown]                       [k] 0xffffffffb4800163
               +0.00%  node_children-29d7c4770dadb48b  [.] __rdl_alloc
     0.02%     +0.00%  node_children-29d7c4770dadb48b  [.] <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::next
     0.02%     +0.00%  libc-2.31.so                    [.] _int_free
               +0.00%  node_children-29d7c4770dadb48b  [.] <rayon::iter::map_with::MapInitConsumer<C,INIT,F> as rayon::iter::plumbing::Consumer<T>>::into_folder
     0.01%     +0.00%  libc-2.31.so                    [.] realloc
               +0.00%  node_children-29d7c4770dadb48b  [.] __rdl_realloc
               +0.00%  node_children-29d7c4770dadb48b  [.] __rdl_dealloc
               +0.00%  node_children-29d7c4770dadb48b  [.] alloc::fmt::format
               +0.00%  node_children-29d7c4770dadb48b  [.] <alloc::vec::Vec<T> as alloc::vec::SpecExtend<&T,core::slice::Iter<T>>>::spec_extend
     0.00%     +0.00%  node_children-29d7c4770dadb48b  [.] plotters::drawing::backend_impl::svg::SVGBackend::escape_and_push
               +0.00%  node_children-29d7c4770dadb48b  [.] __rust_alloc
               +0.00%  node_children-29d7c4770dadb48b  [.] __rust_dealloc
     0.00%     +0.00%  node_children-29d7c4770dadb48b  [.] rayon_core::registry::WorkerThread::steal
     0.00%     +0.00%  libpthread-2.31.so              [.] pthread_cond_wait@@GLIBC_2.3.2
               +0.00%  node_children-29d7c4770dadb48b  [.] rayon_core::sleep::Sleep::get_sleepy
               +0.00%  node_children-29d7c4770dadb48b  [.] crossbeam_epoch::internal::Global::try_advance
               +0.00%  node_children-29d7c4770dadb48b  [.] rayon_core::job::StackJob<L,F,R>::run_inline
               +0.00%  node_children-29d7c4770dadb48b  [.] <plotters::coord::numeric::RangedCoordf64 as plotters::coord::ranged::Ranged>::map
               +0.00%  node_children-29d7c4770dadb48b  [.] <criterion::measurement::DurationFormatter as criterion::measurement::ValueFormatter>::scale_values
               +0.00%  node_children-29d7c4770dadb48b  [.] core::fmt::Formatter::pad_integral
     0.02%     -0.00%  libc-2.31.so                    [.] _int_malloc
     0.00%     +0.00%  node_children-29d7c4770dadb48b  [.] core::fmt::write
               +0.00%  node_children-29d7c4770dadb48b  [.] rayon_core::sleep::Sleep::tickle
               +0.00%  node_children-29d7c4770dadb48b  [.] rayon_core::registry::WORKER_THREAD_STATE::__getit
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] <rayon::iter::map_with::MapWithFolder<C,U,F> as rayon::iter::plumbing::Folder<T>>::complete
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] <rayon::iter::fold::FoldFolder<C,ID,F> as rayon::iter::plumbing::Folder<T>>::complete
     0.00%     +0.00%  node_children-29d7c4770dadb48b  [.] rayon_core::sleep::Sleep::sleep
               +0.00%  node_children-29d7c4770dadb48b  [.] <&mut W as core::fmt::Write>::write_str
     0.00%     -0.00%  libc-2.31.so                    [.] _int_realloc
               +0.00%  node_children-29d7c4770dadb48b  [.] <(alloc::vec::Vec<A>, alloc::vec::Vec<B>, alloc::vec::Vec<C>, alloc::vec::Vec<D>) as criterion::stats::tuple::TupledDistributionsBuilder>::new
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] <rayon::iter::fold::FoldFolder<C,ID,F> as rayon::iter::plumbing::Folder<T>>::consume_iter
               +0.00%  node_children-29d7c4770dadb48b  [.] std::thread::yield_now
               +0.00%  node_children-29d7c4770dadb48b  [.] __rust_probestack
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] criterion::plot::plotters_backend::distributions::abs_distributions
     0.00%     +0.00%  libc-2.31.so                    [.] unlink_chunk.isra.0
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] criterion::analysis::regression
               +0.00%  node_children-29d7c4770dadb48b  [.] <serde::de::impls::<impl serde::de::Deserialize for alloc::vec::Vec<T>>::deserialize::VecVisitor<T> as serde::de::Visitor>::visit_seq
               +0.00%  node_children-29d7c4770dadb48b  [.] rayon_core::sleep::Sleep::tickle_cold
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] crossbeam_epoch::internal::Global::collect
               +0.00%  node_children-29d7c4770dadb48b  [.] <rayon::iter::reduce::ReduceConsumer<R,ID> as rayon::iter::plumbing::Reducer<T>>::reduce
               +0.00%  node_children-29d7c4770dadb48b  [.] oorandom::Rand64::new
               +0.00%  node_children-29d7c4770dadb48b  [.] oorandom::Rand64::rand_u64
     0.00%     +0.00%  node_children-29d7c4770dadb48b  [.] <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
               +0.00%  node_children-29d7c4770dadb48b  [.] core::fmt::num::imp::<impl core::fmt::Display for i32>::fmt
               +0.00%  node_children-29d7c4770dadb48b  [.] __rust_realloc
               +0.00%  node_children-29d7c4770dadb48b  [.] <plotters::coord::numeric::RangedCoordf64 as plotters::coord::ranged::Ranged>::key_points
               +0.00%  node_children-29d7c4770dadb48b  [.] core::fmt::float::float_to_decimal_common_shortest
               +0.00%  libm-2.31.so                    [.] roundf32x
               +0.00%  node_children-29d7c4770dadb48b  [.] <T as plotters::element::dynelem::IntoDynElement<DB,Coord>>::into_dyn
               +0.00%  node_children-29d7c4770dadb48b  [.] <plotters::drawing::backend_impl::svg::SVGBackend as plotters::drawing::backend::DrawingBackend>::fill_polygon
     0.00%     -0.00%  libpthread-2.31.so              [.] __pthread_disable_asynccancel
               +0.00%  node_children-29d7c4770dadb48b  [.] criterion::analysis::common
               +0.00%  node_children-29d7c4770dadb48b  [.] plotters::drawing::backend_impl::svg::SVGBackend::open_tag
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] std::thread::local::LocalKey<T>::with
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] <plotters::drawing::backend_impl::svg::SVGBackend as plotters::drawing::backend::DrawingBackend>::draw_text
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] plotters::drawing::area::DrawingArea<DB,CT>::draw
               +0.00%  node_children-29d7c4770dadb48b  [.] crossbeam_epoch::sync::queue::Queue<T>::try_pop_if
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] rayon::iter::collect::special_extend
     0.00%     -0.00%  libc-2.31.so                    [.] malloc_consolidate
     0.00%     -0.00%  libpthread-2.31.so              [.] pthread_cond_broadcast@@GLIBC_2.3.2
     0.00%     -0.00%  libpthread-2.31.so              [.] __pthread_mutex_lock
     0.00%     +0.00%  libpthread-2.31.so              [.] __pthread_mutex_unlock_usercnt
               +0.00%  node_children-29d7c4770dadb48b  [.] criterion::measurement::ValueFormatter::format_value
     0.00%     +0.00%  node_children-29d7c4770dadb48b  [.] <rayon_core::job::StackJob<L,F,R> as rayon_core::job::Job>::execute
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] rayon_core::registry::in_worker
               +0.00%  node_children-29d7c4770dadb48b  [.] core::ops::function::impls::<impl core::ops::function::FnMut<A> for &F>::call_mut
               +0.00%  node_children-29d7c4770dadb48b  [.] std::sys_common::mutex::raw
               +0.00%  node_children-29d7c4770dadb48b  [.] <sorbus::green::builder::ThinEqNode as core::hash::Hash>::hash
               +0.00%  node_children-29d7c4770dadb48b  [.] <rayon::range::IterProducer<usize> as rayon::iter::plumbing::Producer>::split_at
     0.00%     +0.00%  libpthread-2.31.so              [.] __lll_lock_wait
               +0.00%  node_children-29d7c4770dadb48b  [.] core::fmt::Formatter::pad_integral::write_prefix
               +0.00%  node_children-29d7c4770dadb48b  [.] clap::app::parser::Parser::add_arg
               +0.00%  node_children-29d7c4770dadb48b  [.] plotters::chart::context::ChartContext<DB,plotters::coord::ranged::RangedCoord<X,Y>>::draw_axis_and_labels
               +0.00%  node_children-29d7c4770dadb48b  [.] <std::ffi::os_str::OsStr as core::convert::AsRef<std::ffi::os_str::OsStr>>::as_ref
               +0.00%  node_children-29d7c4770dadb48b  [.] rayon_core::sleep::Sleep::still_sleepy
               +0.00%  libc-2.31.so                    [.] __strlen_avx2
               +0.00%  node_children-29d7c4770dadb48b  [.] plotters::data::float::pretty_print_float
               +0.00%  node_children-29d7c4770dadb48b  [.] core::fmt::Formatter::write_formatted_parts
               +0.00%  node_children-29d7c4770dadb48b  [.] core::fmt::float::float_to_decimal_common_exact
     0.02%     -0.00%  libc-2.31.so                    [.] malloc
     0.00%     -0.00%  libc-2.31.so                    [.] __sched_yield
               +0.00%  node_children-29d7c4770dadb48b  [.] csv::writer::Writer<W>::write_field
               +0.00%  node_children-29d7c4770dadb48b  [.] <plotters::element::text::MultiLineText<Coord,T> as plotters::element::Drawable<DB>>::draw
               +0.00%  node_children-29d7c4770dadb48b  [.] core::unicode::unicode_data::conversions::to_lower
               +0.00%  libm-2.31.so                    [.] __log10_finite
               +0.00%  node_children-29d7c4770dadb48b  [.] plotters::drawing::backend_impl::svg::SVGBackend::init_svg_file
               +0.00%  node_children-29d7c4770dadb48b  [.] plotters::chart::series::SeriesLabelStyle<DB,CT>::draw
               +0.00%  node_children-29d7c4770dadb48b  [.] <plotters::drawing::backend_impl::svg::SVGBackend as plotters::drawing::backend::DrawingBackend>::draw_circle
               +0.00%  node_children-29d7c4770dadb48b  [.] plotters::chart::context::ChartContext<DB,plotters::coord::ranged::RangedCoord<X,Y>>::configure_mesh
     0.00%     +0.00%  libpthread-2.31.so              [.] __pthread_mutex_cond_lock
     0.00%     +0.00%  node_children-29d7c4770dadb48b  [.] ryu::d2s::d2d
     0.00%     +0.00%  node_children-29d7c4770dadb48b  [.] core::num::flt2dec::strategy::grisu::format_shortest_opt
               +0.00%  node_children-29d7c4770dadb48b  [.] <sorbus::green::node::Children as core::iter::traits::iterator::Iterator>::next
               +0.00%  node_children-29d7c4770dadb48b  [.] std::panicking::panicking
               +0.00%  node_children-29d7c4770dadb48b  [.] tinytemplate::compiler::TemplateCompiler::compile
               +0.00%  node_children-29d7c4770dadb48b  [.] clap::app::validator::Validator::validate
               +0.00%  ld-2.31.so                      [.] _dl_lookup_symbol_x
               +0.00%  ld-2.31.so                      [.] _dl_relocate_object
               +0.00%  node_children-29d7c4770dadb48b  [.] tinytemplate::compiler::TemplateCompiler::consume_tag
               +0.00%  libc-2.31.so                    [.] __strchrnul_avx2
               +0.00%  node_children-29d7c4770dadb48b  [.] std::sys::unix::stack_overflow::imp::make_handler
               +0.00%  ld-2.31.so                      [.] _dl_process_pt_note
     0.00%     +0.00%  libc-2.31.so                    [.] new_heap
               +0.00%  node_children-29d7c4770dadb48b  [.] criterion::fs::save_string
     0.00%     +0.00%  libc-2.31.so                    [.] __sysconf
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] criterion::stats::rand_util::SEED_RAND::__getit
               +0.00%  libpthread-2.31.so              [.] __open64
     0.00%     -0.00%  libc-2.31.so                    [.] __clone
     0.00%     -0.00%  libc-2.31.so                    [.] _dl_addr
     0.00%     +0.00%  ld-2.31.so                      [.] __GI___tunables_init
     0.00%     -0.00%  node_children-29d7c4770dadb48b  [.] <plotters::drawing::backend_impl::svg::SVGBackend as plotters::drawing::backend::DrawingBackend>::draw_path
               +0.00%  libc-2.31.so                    [.] __spawni_child
               +0.00%  libpthread-2.31.so              [.] _pthread_cleanup_pop
               +0.00%  libpthread-2.31.so              [.] __condvar_confirm_wakeup
               +0.00%  libc-2.31.so                    [.] __ctype_init
               +0.00%  libpthread-2.31.so              [.] __libc_write
     0.00%     -0.00%  libpthread-2.31.so              [.] __condvar_dec_grefs
    22.75%             node_children-29d7c4770dadb48b  [.] criterion::fs::load
    21.48%             node_children-29d7c4770dadb48b  [.] core::ptr::drop_in_place
    16.09%             node_children-29d7c4770dadb48b  [.] core::hash::impls::<impl core::hash::Hash for (A, B)>::hash
    10.89%             node_children-29d7c4770dadb48b  [.] sorbus::green::node::Node::layout
     4.15%             node_children-29d7c4770dadb48b  [.] sorbus::green::node::Node::new
     3.97%             node_children-29d7c4770dadb48b  [.] hashbrown::raw::RawTable<T>::reserve_rehash
     0.71%             node_children-29d7c4770dadb48b  [.] alloc::collections::btree::node::Handle<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Mut,K,V,alloc::collections::btree::node::marker::Internal>,alloc::collections::btree::node::marker::Edge>::insert
     0.18%             node_children-29d7c4770dadb48b  [.] <criterion::plot::plotters_backend::PlottersBackend as criterion::plot::Plotter>::violin
     0.16%             node_children-29d7c4770dadb48b  [.] criterion::analysis::estimates
     0.15%             node_children-29d7c4770dadb48b  [.] criterion::plot::gnuplot_backend::pdf::pdf_comparison
     0.11%             node_children-29d7c4770dadb48b  [.] <criterion::plot::plotters_backend::PlottersBackend as criterion::plot::Plotter>::t_test
     0.04%             node_children-29d7c4770dadb48b  [.] 0x0000000000032ace
     0.04%             node_children-29d7c4770dadb48b  [.] <textwrap::splitting::HyphenSplitter as textwrap::splitting::WordSplitter>::split
     0.04%             node_children-29d7c4770dadb48b  [.] crossbeam_deque::Worker<T>::resize
     0.01%             node_children-29d7c4770dadb48b  [.] criterion::plot::gnuplot_backend::pdf::pdf_comparison_small
     0.01%             node_children-29d7c4770dadb48b  [.] crossbeam_epoch::collector::Collector::new
     0.01%             node_children-29d7c4770dadb48b  [.] std::sync::mutex::Mutex<T>::new
     0.00%             node_children-29d7c4770dadb48b  [.] <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_struct
     0.00%             node_children-29d7c4770dadb48b  [.] rayon_core::registry::Registry::new
     0.00%             node_children-29d7c4770dadb48b  [.] std::sys_common::util::abort
     0.00%             node_children-29d7c4770dadb48b  [.] crossbeam_epoch::internal::Local::finalize
     0.00%             node_children-29d7c4770dadb48b  [.] <(alloc::vec::Vec<A>,) as criterion::stats::tuple::TupledDistributionsBuilder>::complete
     0.00%             node_children-29d7c4770dadb48b  [.] criterion::routine::Routine::profile
     0.00%             node_children-29d7c4770dadb48b  [.] plotters::chart::context::ChartContext<DB,plotters::coord::ranged::RangedCoord<X,Y>>::draw_mesh
     0.00%             node_children-29d7c4770dadb48b  [.] <criterion::measurement::DurationFormatter as criterion::measurement::ValueFormatter>::scale_throughputs
     0.00%             node_children-29d7c4770dadb48b  [.] alloc::alloc::box_free
     0.00%             node_children-29d7c4770dadb48b  [.] alloc::string::String::push
     0.00%             node_children-29d7c4770dadb48b  [.] rayon_core::registry::WorkerThread::execute
     0.00%             node_children-29d7c4770dadb48b  [.] crossbeam_queue::seg_queue::SegQueue<T>::push
     0.00%             node_children-29d7c4770dadb48b  [.] rayon_core::registry::ThreadInfo::new
     0.00%             node_children-29d7c4770dadb48b  [.] std::sys_common::util::report_overflow
     0.00%             node_children-29d7c4770dadb48b  [.] <(alloc::vec::Vec<A>, alloc::vec::Vec<B>, alloc::vec::Vec<C>, alloc::vec::Vec<D>) as criterion::stats::tuple::TupledDistributionsBuilder>::complete
     0.00%             node_children-29d7c4770dadb48b  [.] <&mut W as core::fmt::Write>::write_char
     0.00%             node_children-29d7c4770dadb48b  [.] crossbeam_epoch::deferred::Deferred::new::call
     0.00%             node_children-29d7c4770dadb48b  [.] std::thread::local::fast::Key<T>::try_initialize
     0.00%             node_children-29d7c4770dadb48b  [.] core::str::traits::_$LT$impl$u20$core..slice..SliceIndex$LT$str$GT$$u20$for$u20$core..ops..range..RangeFrom$LT$usize$GT$$GT$::index::_$u7b$$u7b$closure$u7d$$u7d$::h7b14ff93e6ab52c7
     0.00%             node_children-29d7c4770dadb48b  [.] <alloc::borrow::Cow<str> as core::ops::arith::AddAssign<&str>>::add_assign
     0.00%             node_children-29d7c4770dadb48b  [.] std::process::Child::wait_with_output
     0.00%             node_children-29d7c4770dadb48b  [.] <alloc::boxed::Box<S> as slice_dst::AllocSliceDst<S>>::new_slice_dst
     0.00%             node_children-29d7c4770dadb48b  [.] <serde_json::value::ser::SerializeMap as serde::ser::SerializeStruct>::serialize_field
     0.00%             node_children-29d7c4770dadb48b  [.] <plotters::series::line_series::LineSeries<DB,Coord> as core::iter::traits::iterator::Iterator>::next
     0.00%             node_children-29d7c4770dadb48b  [.] alloc::raw_vec::RawVec$LT$T$C$A$GT$::allocate_in::_$u7b$$u7b$closure$u7d$$u7d$::h3cc6dcb9040591e1
     0.00%             node_children-29d7c4770dadb48b  [.] core::fmt::num::imp::<impl core::fmt::Display for i16>::fmt
     0.00%             node_children-29d7c4770dadb48b  [.] serde_json::value::to_value
     0.00%             node_children-29d7c4770dadb48b  [.] core::fmt::Formatter::pad_formatted_parts
     0.00%             node_children-29d7c4770dadb48b  [.] <crossbeam_epoch::internal::Bag as core::default::Default>::default
     0.00%             node_children-29d7c4770dadb48b  [.] core::unicode::unicode_data::white_space::lookup
     0.00%             node_children-29d7c4770dadb48b  [.] core::fmt::Write::write_char
     0.00%             node_children-29d7c4770dadb48b  [.] core::hash::impls::<impl core::hash::Hash for &T>::hash
     0.00%             node_children-29d7c4770dadb48b  [.] <core::iter::adapters::ResultShunt<I,E> as core::iter::traits::iterator::Iterator>::next
     0.00%             node_children-29d7c4770dadb48b  [.] criterion::analysis::copy_new_dir_to_base
     0.00%             node_children-29d7c4770dadb48b  [.] <criterion::plot::plotters_backend::PlottersBackend as criterion::plot::Plotter>::wait
     0.00%             node_children-29d7c4770dadb48b  [.] std::sys_common::backtrace::_print_fmt::_$u7b$$u7b$closure$u7d$$u7d$::hbec084412a8cd20b
     0.00%             node_children-29d7c4770dadb48b  [.] <plotters::drawing::backend_impl::svg::SVGBackend as plotters::drawing::backend::DrawingBackend>::present
     0.00%             node_children-29d7c4770dadb48b  [.] <&mut W as core::fmt::Write>::write_fmt
     0.00%             node_children-29d7c4770dadb48b  [.] __udivmodti4
     0.00%             node_children-29d7c4770dadb48b  [.] <plotters::drawing::backend::DrawingErrorKind<E> as core::fmt::Debug>::fmt
     0.00%             node_children-29d7c4770dadb48b  [.] plotters::series::area_series::AreaSeries<DB,X,Y>::new
     0.00%             node_children-29d7c4770dadb48b  [.] plotters::coord::ranged::RangedCoord<X,Y>::draw_mesh
     0.00%             node_children-29d7c4770dadb48b  [.] <alloc::collections::btree::map::BTreeMap<K,V> as core::ops::drop::Drop>::drop
     0.00%             node_children-29d7c4770dadb48b  [.] rayon_core::registry::Registry::pop_injected_job
     0.00%             node_children-29d7c4770dadb48b  [.] ryu::pretty::mantissa::write_mantissa_long
     0.00%             node_children-29d7c4770dadb48b  [.] <&plotters::style::text::TextStyle as core::convert::Into<plotters::style::text::TextStyle>>::into
     0.00%             node_children-29d7c4770dadb48b  [.] criterion::plot::plotters_backend::regression::regression_figure
     0.00%             node_children-29d7c4770dadb48b  [.] <plotters::style::font::naive::FontError as core::fmt::Debug>::fmt
     0.00%             node_children-29d7c4770dadb48b  [.] core::fmt::num::<impl core::fmt::Debug for usize>::fmt
     0.00%             node_children-29d7c4770dadb48b  [.] core::num::flt2dec::strategy::dragon::format_exact
     0.00%             node_children-29d7c4770dadb48b  [.] core::fmt::builders::DebugInner::entry
     0.00%             node_children-29d7c4770dadb48b  [.] alloc::sync::Arc<T>::drop_slow
     0.00%             node_children-29d7c4770dadb48b  [.] std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h6ea801765a31224f
     0.00%             node_children-29d7c4770dadb48b  [.] plotters::style::text::TextStyle::transform
     0.00%             node_children-29d7c4770dadb48b  [.] criterion_plot::Figure::save
     0.00%             node_children-29d7c4770dadb48b  [.] rayon_core::registry::Registry::inject
     0.00%             ld-2.31.so                      [.] do_lookup_x
     0.00%             node_children-29d7c4770dadb48b  [.] core::str::slice_error_fail
     0.00%             ld-2.31.so                      [.] strcmp
     0.00%             libc-2.31.so                    [.] malloc_init_state
     0.00%             libpthread-2.31.so              [.] pthread_getattr_np
     0.00%             ld-2.31.so                      [.] _dl_map_object_deps
     0.00%             node_children-29d7c4770dadb48b  [.] std::sys::unix::stack_overflow::imp::signal_handler
```

Note that I benchmarked the whole thing, including the `criterion` harness. I'm not sure how to do it better.

CAD97 commented:
Yikes, you have an even bigger difference for the visit benchmark.

What I've been trying to do is a custom `main` that runs just one of the benchmarking operations (`tree.children().for_each(black_hole)` or `tree.children().for_each(visit)`) some number of thousands of times.

CAD97 commented:
I managed to get Visual Studio's profiler (somewhat) working for an arbitrary exe.

Pre-patch flat:
![image](https://user-images.githubusercontent.com/5992217/79148927-4edd9e00-7d94-11ea-8916-eae1dc64e096.png)

Post-patch flat:
![image](https://user-images.githubusercontent.com/5992217/79148959-5e5ce700-7d94-11ea-9e6f-5374e88fb66e.png)

Pre-patch visit:
![image](https://user-images.githubusercontent.com/5992217/79149871-ceb83800-7d95-11ea-8f24-1a7df162e523.png)

Post-patch visit: 
![image](https://user-images.githubusercontent.com/5992217/79149426-2609d880-7d95-11ea-9e5d-bf5d2bd68997.png)

I still really don't know how to interpret any of these results, but the fact that visit is longer-running and sees such a large, consistent improvement, suggests to me that doing this (manually tracking alignment parity) is the correct thing to do. Maybe someone with better knowledge on how to microbenchmark and tune things like this can touch it in the future; right now, I really don't know how to go about doing it.

CAD97 committed:
Teach node children iterator about element parity

NOTE: This is a performance, not correctness patch.

See discussion in the PR:
https://github.com/CAD97/sorbus/pull/1#issuecomment-612549153

On my machine, the flat iteration took an approximately 9% regression in time,
and about an 8% regression in throughput. On the other hand, the visitor saw
a 33% reduction in time, or a 50% improvement in throughput!

Inicola's machine (linux) saw a consistently directed but better result, with
approx. a 3/3% regression on flat iteration and a 43/75% (!!!) improvement for
the visitor benchmark.

Because the visitor benchmark is the longer-running and saw a large benefit,
in comparision to the smaller penalty on the flat iteration, along with the
assumption that visit is closer to a realistic workload, I have decided to
include this patch. It also feels "morally correct" to teach the children
iterator about the parity of the child elements, rather than checking the
alignment every single time.

As a Windows user with little performance measurement and tuning experience,
and with not-so-useful results from actually attempting to measure performance,
(surprise: all of the time is spent in `Children::fold`), I (CAD97) cannot do
much more in tuning here. I welcome those with more experience micro-optimizing
to take a stab at tuning iteration here, though to be frank, I do not expect
iteration of green syntax tree nodes to be a significant factor in performance
of any real application; the domain logic will likely easily outweigh it. Then,
the importance is more about how well the optimizer optimizes through the loop
than the actual operations and speed of the loop itself.
